agent/bench-scripts/postprocess/process-iteration-samples

```diff
@@ -220,24 +227,33 @@ foreach $sample (@json_samples) { # samples 0..N
 						if ($key eq get_label('value_label')) {
 							my %samples_hash;
 							$samples_hash{get_label('value_label')} = $sample_value;
-							if ($sample_hash{$metric_type}{$metric_name}[$i]{get_label('timeseries_label')}) {
+							if (exists($sample_hash{$metric_type}{$metric_name}[$i]{get_label('timeseries_label')})) {
 								$samples_hash{get_label('timeseries_label')} = $sample_hash{$metric_type}{$metric_name}[$i]{get_label('timeseries_label')};
 							}
-							if ( not defined $iteration{$metric_type}{$metric_name}[$iteration_matching_index]{get_label('samples_label')} ) {
+							if (exists($iteration{$metric_type}{$metric_name}[$iteration_matching_index]{get_label('samples_label')})) {
 								# Create an array to hold the samples for "value" values
 								# For example, if uperf had 5 sample runs, there would be an array of 5 Gb_sec values in this array
-								#my @samples;
 								$iteration{$metric_type}{$metric_name}[$iteration_matching_index]{get_label('samples_label')} = ();
 							}
 							push @{$iteration{$metric_type}{$metric_name}[$iteration_matching_index]{get_label('samples_label')}}, \%samples_hash;
-						# For the rest of the keys, either check to see if it matches existing value in the %iteration, 
-						# or if it does not exist, create it.
 						} else {
+							# For the rest of the keys, either check to see if it matches existing value in the %iteration,
+							# or if it does not exist, create it.
 							if ( $key ne get_label('timeseries_label') and $key ne get_label('uid_label') ) { # timeseries already handled above
-								if ($iteration{$metric_type}{$metric_name}[$iteration_matching_index]{$key}) {
+								if (exists($iteration{$metric_type}{$metric_name}[$iteration_matching_index]{$key})) {
 									my $iteration_value = $iteration{$metric_type}{$metric_name}[$iteration_matching_index]{$key};
 									if ("$sample_value" ne "$iteration_value") {
-										print "Error: values do not match\nkey: $key values: $iteration_value $sample_value\n";
+										my $sv = ref($sample_value);
+										my $iv = ref($iteration_value);
+										if ($sv and ($sv eq $iv)) {
+											my $sv_s = Dumper($sample_value);
+											my $iv_s = Dumper($iteration_value);
+											if ($sv_s ne $iv_s) {
+												print "Error: values do not match\n\tkey: $key\n\titeration value: $iv_s\n\tsample    value: $sv_s\n";
+											}
+										} else {
+											print "Error: values do not match\n\tkey: $key\n\titeration value: $iteration_value\n\tsample    value: $sample_value\n";
+										}
```

Comment:
I am __so__ confused. Trying to read Perl gives me such a headache.

So, if a Perl comparison of `$sample_value` and `$iteration_value` shows they're _not_ equal, we get `ref()` of both to see whether they are refs to the same type; if so, then we `Dumper` both to get a deep representation of whatever they reference and compare _those_; we fail if the `Dumper`s don't match, and we fail if they're not identical types in the first place.

Why do they have to be identical objects just because they're the same type? And a better error message for the `else` would be that they're not the same type rather than repeating that the values don't match.

Reply:
> > I am **so** confused.
> 
> Based on what you wrote below, you do not appear to be confused. =)

I was quite sure I was confused; so if I'm not, that's confusing...

> > Why do they have to be identical objects just because they're the same type?
> 
> In the first place, the original code was emitting error messages when the iteration key and the sample key have different values.

Right.

> In the case where the values are not scalars, they would always emit the error message, but an array and a hash _always_ have different scalar values.

I don't think you explained that clearly, but I get it: two identical hashes or two identical arrays (which don't happen to be refs to the same Perl object) would have appeared different before but now will be diagnosed as equal.

> This change is trying to fix that. If each key is referencing the same type, and the `Dumper` of each are identical, then we consider them to have the same data, and we don't emit an error message.
> 
> If they are not the same type, they certainly can't be the same values, so we just emit the same message as before.
> 
> > And a better error message for the `else` would be that they're not the same type rather than repeating that the values don't match.
> 
> Well, that doesn't help the user. They don't know anything about Perl. That just want to know what keys are being compared and why the don't match. So this code only emits information about the keys and values.

I always find explanations like this a bit concerning. Basically, you _are_ telling them that the Perl types don't match, because the implementation is in Perl and the semantics aren't separable. Then again, I can't immediately construct a case where the two strings you dump would "appear equivalent" but be distinct to Perl so I'm not going to pursue that debate; I don't think it really matters, although I thought it might be useful to _us_ when someone shows us the error message.

Reply:
It is pretty easy to see the difference: when you have a hash or an array it is displayed as such, vs a string, float, or integer which don't look like those objects as rendered by `Dumper`.